src/util-interval-tree.c

```diff
@@ -0,0 +1,189 @@
+{
+    DEBUG_VALIDATE_BUG_ON(p->port > p->port2);
+
+    SCIntervalNode *pi = SCCalloc(1, sizeof(*pi));
+    if (pi == NULL) {
+        return -1;
+    }
+    pi->port = p->port;
+    pi->port2 = p->port2;
+    pi->flags = p->flags;
+    SigGroupHeadCopySigs(de_ctx, p->sh, &pi->sh);
+    if (PI_IRB_INSERT(&it->tree, pi) != NULL) {
+        SCLogNotice("Node wasn't added to the tree: port: %d, port2: %d", pi->port, pi->port2);
+        SCFree(pi);
+        return SC_EINVAL;
+    }
+#if 0
+    SCIntervalNode *root = IRB_ROOT(&it->tree);
+    SCLogNotice("Inserted [%d, %d]; %d, ROOT: [%d, %d]; %d", pi->port, pi->port2,
+            IRB_COLOR(pi, irb), root->port, root->port2, IRB_COLOR(root, irb));
+    printIT(IRB_ROOT(&it->tree), 0);
+#endif
+    return SC_OK;
+}
+
+static uint16_t cnt_overlaps = 0;
+
+static bool IsOverlap(uint16_t port, uint16_t port2, SCIntervalNode *ptr)
+{
+    if ((port <= ptr->port2) && (ptr->port <= port2)) {
+        SCLogNotice("Found overlap with [%d, %d]", ptr->port, ptr->port2);
+        cnt_overlaps++;
+        return true;
+    }
+    SCLogDebug("No overlap found for [%d, %d) w [%d, %d]", port, port2, ptr->port, ptr->port2);
+    return false;
+}
+
+static void FindOverlaps(DetectEngineCtx *de_ctx, uint16_t port, uint16_t port2, SCIntervalNode *ptr, DetectPort **list)
+{
+    SCIntervalNode *prev_ptr = NULL;
+    bool is_overlap = false;
+    DetectPort *new_port = NULL;
+    while (ptr != prev_ptr) {
+        prev_ptr = ptr;
+        is_overlap = IsOverlap(port, port2, ptr);
+        if (is_overlap && (cnt_overlaps == 1)) {
+            // Allocate memory for port obj
+            new_port = DetectPortInit();
+            if (new_port == NULL) {
+                goto error;
+            }
+            new_port->port = port;
+            new_port->port2 = port2;
+            SigGroupHeadCopySigs(de_ctx, ptr->sh, &new_port->sh);
+            if (*list == NULL) {
+                *list = new_port;
+                (*list)->last = new_port;
+            } else {
+                (*list)->last->next = new_port;
+                (*list)->last = new_port;
+            }
+        } else if (new_port != NULL && is_overlap && cnt_overlaps > 1) {
+            SigGroupHeadCopySigs(de_ctx, ptr->sh, &new_port->sh);
+        }
+        SCIntervalNode *node = IRB_LEFT(ptr, irb);
+        if ((node != NULL) && (node->max >= port)) {
+            ptr = node;
+        } else {
+            node = IRB_RIGHT(ptr, irb);
+            if ((node != NULL) && (ptr->port < port2) && (node->max >= port)) {
+                ptr = node;
+            }
+        }
+    }
+    cnt_overlaps = 0;
+    return;
+error:
+    if (new_port != NULL)
+        DetectPortFree(de_ctx, new_port);
+    return;
+}
+
+void PISearchOverlappingPortRanges(DetectEngineCtx *de_ctx,
+        uint16_t port, uint16_t port2, struct PI *head, DetectPort **list)
+{
+    SCIntervalNode *ptr = IRB_ROOT(head);
```

Comment:
in theory, ptr could be NULL here and potentially paths exist in `FindOverlaps` that lead to NULL-ptr deref. Maybe not be reachable in practice, but I sleep better knowing things are checked or explained :)